### PR TITLE
Remove go 1.19 support, bump minimum to go 1.20 and add testing for 1.21

### LINKED ISSUE #25116: Add testing for Go 1.21, drop testing for Go 1.19
### Component(s)

_No response_

### Describe the issue you're reporting

Same as #18454 but for Go 1.21/Go 1.19.

Similar to https://github.com/open-telemetry/opentelemetry-collector/issues/8207, we want to remove testing with Go 1.19, add testing with Go 1.21 and allow contributors to use Go 1.20+ specific features.

This is blocked by https://github.com/open-telemetry/opentelemetry-collector/issues/8207

This unblocks the updates listed in #20245

### PR DESCRIPTION
**Pull request by shivanshuraj1333**
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/25116

### CODE REVIEW

**Comment by mx-psi** [comment]
Thanks @shivanshu1333! Note that we are going to merge #25151 first

File: CHANGELOG.md [hunk 1]
```diff
@@ -950,7 +950,7 @@ If you are looking for developer-facing changes, check out [CHANGELOG-API.md](./
 ### 🛑 Breaking changes 🛑
 
 - `fluentbitextension`: remove deprecated extension (#18505)
-- `all`: Remove go 1.18 support, bump minimum to go 1.19 and add testing for 1.20 (#18436)
+- `all`: Remove go 1.18 support, bump minimum to go 1.20 and add testing for 1.20 (#18436)
```
Comment (mx-psi) [inline CHANGELOG.md:953]: This change needs to be reverted

File: examples/demo/client/Dockerfile [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # Copyright The OpenTelemetry Authors
 # SPDX-License-Identifier: Apache-2.0
-FROM golang:1.20 as build
+FROM golang:1.20.7.7 as build
```
Comment (mx-psi) [inline examples/demo/client/Dockerfile:3]:
```suggestion
FROM golang:1.20.7 as build
```

File: cmd/telemetrygen/Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-FROM golang:1.20 AS build
+FROM golang:1.20.7.7 AS build
```
Comment (mx-psi) [inline cmd/telemetrygen/Dockerfile:1]:
```suggestion
FROM golang:1.20.7 AS build
```

File: .chloggen/go1-21.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: breaking
+
+# The name of the component, or a single word describing the area of concern, (e.g. otlpreceiver)
+component: all
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Remove go 1.20 support, bump minimum to go 1.20 and add testing for 1.21
```
Comment (mx-psi) [inline .chloggen/go1-21.yaml:8]:
```suggestion
note: Remove go 1.19 support, bump minimum to go 1.20 and add testing for 1.21
```

File: CHANGELOG.md (hunk 1 above)
Comment (shivanshu1333) [reply CHANGELOG.md]: my bad, done, thanks!

**Comment by dmitryax** [comment]
@shivanshu1333 please rebase

**Comment by shivanshu1333** [comment]
Done!
Is it good to go now?
Approvals/reviews requested...

File: cmd/oteltestbedcol/go.mod [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 module github.com/open-telemetry/opentelemetry-collector-contrib/cmd/oteltestbedcol
 
-go 1.19
+go 1.20
```
Comment (shivanshu1333) [inline cmd/oteltestbedcol/go.mod:5]: lint is failing on this, running `make genotelcontribcol` reverts this change to go 1.19
How to fix linting here?

Comment (dmitryax) [reply cmd/oteltestbedcol/go.mod:5]: I'd suggest looking at the same PR done for the 1.19 and see what's missed

Comment (shivanshu1333) [reply cmd/oteltestbedcol/go.mod:5]: Problem solved, had to bump `go.opentelemetry.io/collector/cmd/builder` to latest sha of master.

File: cmd/telemetrygen/Dockerfile [hunk 6]
```diff
@@ -1,4 +1,4 @@
-FROM golang:1.20 AS build
+FROM golang:1.20.7 AS build
```
Comment (dmitryax) [inline cmd/telemetrygen/Dockerfile:1]: This is not needed

File: examples/demo/client/Dockerfile [hunk 7]
```diff
@@ -1,6 +1,6 @@
 # Copyright The OpenTelemetry Authors
 # SPDX-License-Identifier: Apache-2.0
-FROM golang:1.20 as build
+FROM golang:1.20.7 as build
```
Comment (shivanshu1333) [inline examples/demo/client/Dockerfile:3]: @dmitryax should I also revert this and other such changes?

File: cmd/telemetrygen/Dockerfile (hunk 6 above)
Comment (shivanshu1333) [reply cmd/telemetrygen/Dockerfile]: I think the correct conversion here is to change it to 1.21
ref: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18436/files?file-filters%5B%5D=No+extension&show-viewed-files=true#diff-15a455599a44216f0cc2d8d039375699ab5408b90dad03fff2f5040bcadb3f84L1

File: examples/demo/client/Dockerfile (hunk 7 above)
Comment (shivanshu1333) [reply examples/demo/client/Dockerfile:3]: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/25162#discussion_r1292834345

File: cmd/telemetrygen/Dockerfile (hunk 6 above)
Comment (shivanshu1333) [reply cmd/telemetrygen/Dockerfile]: looking at older [PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18436/files?file-filters%5B%5D=No+extension&show-viewed-files=true#diff-15a455599a44216f0cc2d8d039375699ab5408b90dad03fff2f5040bcadb3f84L1)
Yes, this is not needed, I think all the `Dockerfile` changes can be reverted, added a commit.

File: examples/demo/client/Dockerfile (hunk 7 above)
Comment (shivanshu1333) [reply examples/demo/client/Dockerfile]: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/25162#discussion_r1292834906

**Comment by shivanshu1333** [comment]
I think this is ready for merge, lmk if anything else is needed :)

File: examples/demo/client/Dockerfile (hunk 7 above)
Comment (dmitryax) [reply examples/demo/client/Dockerfile]: Yes please. We should not define patch version of the docker image

**Review by dmitryax — APPROVED** [review]

**Pull request merged**